### PR TITLE
[plugin.program.AML] 0.10.0

### PR DESCRIPTION
**Pull request by Wintermute0110**
### Description

AML 0.10.0 is the next version of the addon for Krypton and Leia. The difference with the previous version is huge due to a refactoring. This refactoring was part of the process when upgrading the addon to Python 3. AML has 20k something lines of Python code and after the refactoring the differences between the Python 2 and Python 3 versions is less than 50 lines, not using Python six, which I think is quite remarkable. The number of new features is small compared with the previous version of the addon, for more details see the `changelog.txt` file.

### Checklist:

- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

File: plugin.program.AML/addon.xml [hunk 1]
```diff
@@ -1,44 +1,48 @@
...
+        <platform>android freebsd ios linux osx windows</platform>
+        <license>GPL-2.0-only</license>
+        <email>wintermute0110@gmail.com</email>
+        <forum>https://forum.kodi.tv/showthread.php?tid=304186</forum>
+        <website>https://github.com/Wintermute0110/plugin.program.AML.dev/wiki</website>
+        <source>https://github.com/Wintermute0110/plugin.program.AML.dev/</source>
+        <assets>
+            <icon>media/icon.png</icon>
+            <fanart>media/fanart.jpg</fanart>
+            <screenshot>media/shot_01_main_window.jpg</screenshot>
```
Comment (enen92) [inline plugin.program.AML/addon.xml:34]: can you please limit screenshots to 10?

File: plugin.program.AML/resources/db.py [hunk 2]
```diff
@@ -0,0 +1,1228 @@
...
+# Hash database with 256 elements (2 hex digits)
+def db_build_main_hashed_db(cfg, control_dic, machines, machines_render):
+    log_info('db_build_main_hashed_db() Building main hashed database...')
+
+    # machine_name -> MD5 -> take two letters -> aa.json, ab.json, ...
+    # A) First create an index
+    #    db_main_hash_idx = { 'machine_name' : 'aa', ... }
+    # B) Then traverse a list [0, 1, ..., f] and write the machines in that sub database section.
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress('Building main hashed database...', len(machines))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:936]: has to be localized

File: plugin.program.AML/resources/db.py [hunk 3]
```diff
@@ -0,0 +1,1228 @@
...
+        db_main_hash_idx[key] = db_name
+        # log_debug('Machine {:20s} / hash {} / db file {}'.format(key, md5_str, db_name))
+    pDialog.endProgress()
+
+    hex_digits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
+    distributed_db_files = []
+    for u in range(len(hex_digits)):
+        for v in range(len(hex_digits)):
+            distributed_db_files.append('{}{}'.format(hex_digits[u], hex_digits[v]))
+    pDialog.startProgress('Building main hashed database JSON files...', len(distributed_db_files))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:951]: same

File: plugin.program.AML/resources/db.py [hunk 4]
```diff
@@ -0,0 +1,1228 @@
...
+
+    return hashed_db_dic[machine_name]
+
+# MAME hash database with 256 elements (2 hex digits)
+def db_build_asset_hashed_db(cfg, control_dic, assets_dic):
+    log_info('db_build_asset_hashed_db() Building assets hashed database ...')
+
+    # machine_name -> MD5 -> take two letters -> aa.json, ab.json, ...
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress('Building asset hashed database...', len(assets_dic))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:991]: same

File: plugin.program.AML/resources/db.py [hunk 5]
```diff
@@ -0,0 +1,1228 @@
...
+        db_main_hash_idx[key] = db_name
+        # log_debug('Machine {:20s} / hash {} / db file {}'.format(key, md5_str, db_name))
+    pDialog.endProgress()
+
+    hex_digits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
+    distributed_db_files = []
+    for u in range(len(hex_digits)):
+        for v in range(len(hex_digits)):
+            distributed_db_files.append('{}{}'.format(hex_digits[u], hex_digits[v]))
+    pDialog.startProgress('Building asset hashed database JSON files...', len(distributed_db_files))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1006]: same

File: plugin.program.AML/resources/db.py [hunk 6]
```diff
@@ -0,0 +1,1228 @@
...
+    # Notify user this is a forced build.
+    if not cfg.settings['debug_enable_MAME_render_cache'] and force_build:
+        t = 'MAME render cache disabled but forcing rebuilding.'
+        log_info(t)
+        kodi_dialog_OK(t)
+
+    # --- Clean 'cache' directory JSON ROM files ---
+    log_info('Cleaning dir "{}"'.format(cfg.CACHE_DIR.getPath()))
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress('Listing render cache JSON files...')
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1057]: same

File: plugin.program.AML/resources/db.py [hunk 7]
```diff
@@ -0,0 +1,1228 @@
...
+    file_list = os.listdir(cfg.CACHE_DIR.getPath())
+    log_info('Found {} files'.format(len(file_list)))
+    deleted_items = 0
+    pDialog.resetProgress('Cleaning render cache JSON files...', len(file_list))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1061]: same here

File: plugin.program.AML/resources/db.py [hunk 8]
```diff
@@ -0,0 +1,1228 @@
...
+    log_info('Deleted {} files'.format(deleted_items))
+
+    # --- Build ROM cache ---
+    num_catalogs = len(cache_index_dic)
+    catalog_count = 1
+    pDialog.startProgress('Building MAME render cache')
+    for catalog_name in sorted(cache_index_dic):
+        catalog_index_dic = cache_index_dic[catalog_name]
+        catalog_all = db_get_cataloged_dic_all(cfg, catalog_name)
+        diag_t = 'Building MAME [COLOR orange]{}[/COLOR] render cache ({} of {})...'.format(
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1079]: same

File: plugin.program.AML/resources/db.py [hunk 9]
```diff
@@ -0,0 +1,1228 @@
...
+        # log_debug('UNLINK "{}"'.format(full_path))
+        os.unlink(full_path)
+        deleted_items += 1
+    pDialog.endProgress()
+    log_info('Deleted {} files'.format(deleted_items))
+
+    # --- Build ROM cache ---
+    num_catalogs = len(cache_index_dic)
+    catalog_count = 1
+    pDialog.startProgress('Building MAME render cache')
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1075]: here as well

File: plugin.program.AML/resources/db.py [hunk 10]
```diff
@@ -0,0 +1,1228 @@
...
+    # Notify user this is a forced build.
+    if not cfg.settings['debug_enable_MAME_render_cache'] and force_build:
+        t = 'MAME asset cache disabled but forcing rebuilding.'
+        log_info(t)
+        kodi_dialog_OK(t)
+
+    # --- Clean 'cache' directory JSON Asset files ---
+    log_info('Cleaning dir "{}"'.format(cfg.CACHE_DIR.getPath()))
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress('Listing asset cache JSON files...')
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1126]: localize

File: plugin.program.AML/resources/db.py [hunk 11]
```diff
@@ -0,0 +1,1228 @@
...
+    file_list = os.listdir(cfg.CACHE_DIR.getPath())
+    log_info('Found {} files'.format(len(file_list)))
+    deleted_items = 0
+    pDialog.resetProgress('Cleaning asset cache JSON files...', len(file_list))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1130]: localize

File: plugin.program.AML/resources/db.py [hunk 12]
```diff
@@ -0,0 +1,1228 @@
...
+    log_info('Deleted {} files'.format(deleted_items))
+
+    # --- Build MAME asset cache ---
+    num_catalogs = len(cache_index_dic)
+    catalog_count = 1
+    pDialog.startProgress('Building MAME asset cache')
+    for catalog_name in sorted(cache_index_dic):
+        catalog_index_dic = cache_index_dic[catalog_name]
+        catalog_all = db_get_cataloged_dic_all(cfg, catalog_name)
+        diag_t = 'Building MAME [COLOR orange]{}[/COLOR] asset cache ({} of {})...'.format(
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1148]: localize

File: plugin.program.AML/resources/db.py [hunk 13]
```diff
@@ -0,0 +1,1228 @@
...
+# Load and save a bunch of JSON files
+# -------------------------------------------------------------------------------------------------
+#
+# Accepts a list of JSON files to be loaded. Displays a progress dialog.
+# Returns a dictionary with the context of the loaded files.
+#
+def db_load_files(db_files):
+    log_debug('db_load_files() Loading {} JSON database files...'.format(len(db_files)))
+    db_dic = {}
+    d_text = 'Loading databases...'
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1186]: localize

File: plugin.program.AML/resources/db.py [hunk 14]
```diff
@@ -0,0 +1,1228 @@
...
+        dict_key, db_name, db_path = f_item
+        pDialog.updateProgressInc('{}\nDatabase [COLOR orange]{}[/COLOR]'.format(d_text, db_name))
+        db_dic[dict_key] = utils_load_JSON_file_dic(db_path)
+    pDialog.endProgress()
+
+    return db_dic
+
+def db_save_files(db_files, json_write_func = utils_write_JSON_file):
+    log_debug('db_save_files() Saving {} JSON database files...'.format(len(db_files)))
+    d_text = 'Saving databases...'
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1199]: here too

File: plugin.program.AML/resources/db.py [hunk 15]
```diff
@@ -0,0 +1,1228 @@
...
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress(d_text, len(db_files))
+    for f_item in db_files:
+        dict_data, db_name, db_path = f_item
+        pDialog.updateProgressInc('{}\nDatabase [COLOR orange]{}[/COLOR]'.format(d_text, db_name))
```
Comment (enen92) [inline plugin.program.AML/resources/db.py:1204]: localize

File: plugin.program.AML/resources/graphics.py [hunk 16]
```diff
@@ -672,119 +672,100 @@ def graphs_load_MAME_Fanart_template(Template_FN):
...
     # --- Load Assets DB ---
-    pDialog = xbmcgui.DialogProgress()
-    pDialog.create('Advanced MAME Launcher', 'Loading MAME asset database ... ')
-    pDialog.update(0)
-    assets_dic = fs_load_JSON_file_dic(PATHS.MAIN_ASSETS_DB_PATH.getPath())
-    pDialog.update(100)
-    pDialog.close()
-    data_dic['assets_dic'] = assets_dic
+    pDialog = KodiProgressDialog()
+    pDialog.startProgress('Loading MAME asset database...')
```
Comment (enen92) [inline plugin.program.AML/resources/graphics.py:715]: localize

File: plugin.program.AML/resources/graphics.py [hunk 17]
```diff
@@ -672,119 +672,100 @@ def graphs_load_MAME_Fanart_template(Template_FN):
...
     pDialog_canceled = False
-    pDialog = xbmcgui.DialogProgress()
-    pDialog_line1 = 'Building MAME machine Fanarts ...'
-    pDialog.create('Advanced MAME Launcher', pDialog_line1)
-    total_machines, processed_machines = len(data_dic['assets_dic']), 0
+    pDialog = KodiProgressDialog()
+    total_machines, processed_machines = len(data_dic['assetdb']), 0
     ETA_str = ETA_reset(total_machines)
-    for m_name in sorted(data_dic['assets_dic']):
+    diag_t = 'Building MAME machine Fanarts...'
```
Comment (enen92) [inline plugin.program.AML/resources/graphics.py:730]: localize

File: plugin.program.AML/resources/graphics.py [hunk 18]
```diff
@@ -672,119 +672,100 @@ def graphs_load_MAME_Fanart_template(Template_FN):
...
+    pDialog.startProgress(diag_t, total_machines)
+    for m_name in sorted(data_dic['assetdb']):
         build_time_start = time.time()
-        pDialog.update((processed_machines * 100) // total_machines, pDialog_line1,
-            'ETA {0} machine {1}'.format(ETA_str, m_name))
-        if pDialog.iscanceled():
+        pDialog.updateProgress(processed_machines, '{}\nETA {} machine {}'.format(diag_t, ETA_str, m_name))
```
Comment (enen92) [inline plugin.program.AML/resources/graphics.py:734]: localize

File: plugin.program.AML/resources/graphics.py [hunk 19]
```diff
@@ -672,119 +672,100 @@ def graphs_load_MAME_Fanart_template(Template_FN):
...
-    # --- MAME Fanart build timestamp ---
-    control_dic = fs_load_JSON_file_dic(PATHS.MAIN_CONTROL_PATH.getPath())
-    change_control_dic(control_dic, 't_MAME_fanart_build', time.time())
-    fs_write_JSON_file(PATHS.MAIN_CONTROL_PATH.getPath(), control_dic)
+    pDialog.endProgress()
 
-    # --- assets_dic has changed. Rebuild hashed database ---
-    fs_build_asset_hashed_db(PATHS, settings, control_dic, data_dic['assets_dic'])
+    # Save MAME assets DB
+    pDialog.startProgress('Saving MAME asset database...')
```
Comment (enen92) [inline plugin.program.AML/resources/graphics.py:759]: localize

File: plugin.program.AML/resources/graphics.py [hunk 20]
```diff
@@ -829,226 +810,203 @@ def graphs_load_SL_Fanart_template(Template_FN):
...
     }
 
-    # If artwork directory not configured abort.
-    # SL Fanart directories are created later in graphs_build_SL_Fanart_all()
-    if not settings['assets_path']:
-        kodi_dialog_OK('Asset directory not configured. Aborting Fanart generation.')
-        data_dic['abort'] = True
+    # --- If artwork directory not configured abort ---
+    if not cfg.settings['assets_path']:
+        kodi_set_error_status(st_dic, 'Asset directory not configured. Aborting SL Fanart generation.')
```
Comment (enen92) [inline plugin.program.AML/resources/graphics.py:832]: localize?

**Comment by Wintermute0110** [comment]
@enen92 Are you asking me to localize a 20k lines of code addon? Seriously?

**Comment by enen92** [comment]
Usually it is how it works...strings shown on the ui should be localized. 
However I confess I gave up along the review and thought it will be too much work at this point. Please keep it in mind for future updates or as you further work in the addon code.

**Review by enen92 — APPROVED** [review]

**Comment by Wintermute0110** [comment]
@enen92 Thanks for merging. Currently I do not have time for the localization. The development takes all my free time. If somebody wants to localize the Advanced Launchers I will be happy to merge the changes.

When I forked Advanced Launcher code for the first time I decided to completely get rid of the localization because localized code is extremely difficult to change and modify. Small video/audio addons have few strings to localize but the Advanced Launchers, being complete games databases with hundredths of different lines of text, are another story.

Using string IDs for the localized strings could be OK for small addons but huge addons like the Advanced Launcher would require a localization scheme like `print(_('Hello world'))`, and a automatic tool to search for string calling the function `_()`. Is a functionality like `gettext` available in Kodi OR only ID strings are used?

**Pull request merged**